### PR TITLE
Automatic definitions versioning for nxs file

### LINKED ISSUE #162: Add versioning for nexus_definitions
- [x] Add NeXus_version to NXroot

### PR DESCRIPTION
**Pull request by domna**
This adds an example for writing the nexus version and pynxtools version into an ellipsometry file and NXroot (for every file). It tries to extract the versions from git tags and commits, creating version numbers like `v2023.10-10-g<hash>` (for nexus_definitions) to show that the current state has a commit distance of 10 compared to the `v2023.10` tag and also includes the git hash of that commit. For pynxtools it uses setuptools-scm (i.e., yielding a similar version tag).

- [x] Add git hash in `.../definition/url`
- ~[ ] Add git hash in `/entry/program_name/@url`~
- [x] Add NeXus url to NXroot

Fixes #162

### CODE REVIEW

**Comment by domna** [comment]
https://github.com/FAIRmat-NFDI/nexus_definitions/pull/81 deals with versioning for the nexus definitions. We can just use this from the submodule or integrate our own version of it here.

**Comment by sherjeelshabih** [comment]
> [FAIRmat-NFDI/nexus_definitions#81](https://github.com/FAIRmat-NFDI/nexus_definitions/pull/81) deals with versioning for the nexus definitions. We can just use this from the submodule or integrate our own version of it here.

This is neat. I would vote for having this in the submodule with the definitions though. Seems cleaner to have it next to what it's versioning.

**Comment by domna** [comment]
> > [FAIRmat-NFDI/nexus_definitions#81](https://github.com/FAIRmat-NFDI/nexus_definitions/pull/81) deals with versioning for the nexus definitions. We can just use this from the submodule or integrate our own version of it here.
> 
> This is neat. I would vote for having this in the submodule with the definitions though. Seems cleaner to have it next to what it's versioning.

Yes, this would be my preferred solution, too (+ the implementation here does it like this). Initially, I wasn't sure whether pynxtools would be able to get the function from the module as nexus_definitions are not really a python module, but obviously this works.
It depends, however, on the NIAC accepting this function in their repo. If they don't want it I would prefer to move it in pynxtools for simplicity.

**Comment by domna** [comment]
This is a bit hacky but works. It tries to get the nexus version from git, or from `pynxtools/nexus-version.txt`. If both fail the `get_nxdl_version` from the nexus_definitions is used (w/o git hash). The `nexus_version.txt` is created during package build so it's automatically added and mostly be invisible to users.

This implements everything inside pynxtools, so there is nothing new necessary to be added in nexus_definitions. This has to be this way due to the build process hook. So the question is if we still want to add https://github.com/FAIRmat-NFDI/nexus_definitions/pull/81 and propose this to niac or just leave it out.

File: MANIFEST.in [hunk 1]
```diff
@@ -1,4 +1,6 @@
 recursive-include pynxtools/definitions/base_classes/ *.xml
 recursive-include pynxtools/definitions/applications/ *.xml
 recursive-include pynxtools/definitions/contributed_definitions/ *.xml
-include pynxtools/definitions/ *.xsd
+include pynxtools/definitions/*.xsd
```
Comment (sanbrock) [inline MANIFEST.in:4]: if this line needed to be reformatted, why is not the reformatting of the above lines of *.xml necessary?

File: examples/ellipsometry/eln_data.yaml [hunk 2]
```diff
@@ -58,9 +58,6 @@ colnames:
 - Delta
 - err.Psi
 - err.Delta
-definition: NXellipsometry
```
Comment (sanbrock) [inline examples/ellipsometry/eln_data.yaml:61]: why is it needed to be moved to the code and not to keep it separate in the config file?

File: pynxtools/_build_wrapper.py [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+from typing import Optional
+
+from setuptools import build_meta as _orig
+from setuptools.build_meta import *  # pylint: disable=wildcard-import,unused-wildcard-import
+
+
+def get_vcs_version(tag_match="*[0-9]*") -> Optional[str]:
+    """
+    The version of the Nexus standard and the NeXus Definition language
+    based on git tags and commits
```
Comment (sanbrock) [inline pynxtools/_build_wrapper.py:16]: what about the git repo used?

File: pynxtools/dataconverter/helpers.py [hunk 4]
```diff
@@ -591,7 +592,7 @@ def add_default_root_attributes(data, filename):
     data["/@file_name"] = filename
     data["/@file_time"] = str(datetime.now(timezone.utc).astimezone())
     data["/@file_update_time"] = data["/@file_time"]
-    data["/@NeXus_version"] = "NOT_IMPLEMENTED"
+    data["/@NeXus_version"] = get_nexus_version()
```
Comment (sanbrock) [inline pynxtools/dataconverter/helpers.py:595]: is not it overwritten by writer?

File: pynxtools/dataconverter/readers/ellips/reader.py [hunk 5]
```diff
@@ -466,6 +467,16 @@ def read(self,
         for index in range(1, len(data_list)):
             template["/ENTRY[entry]/plot/@auxiliary_signals"] += data_list[index]
 
+        template["/ENTRY[entry]/definition"] = "NXellipsometry"
+        template["/ENTRY[entry]/definition/@url"] = (
+            "https://github.com/FAIRmat-NFDI/nexus_definitions/"
+            "blob/fairmat/contributed_definitions/NXellipsometry.nxdl.xml"
+        )
+        template["/ENTRY[entry]/definition/@version"] = get_nexus_version()
```
Comment (sanbrock) [inline pynxtools/dataconverter/readers/ellips/reader.py:475]: should not it be put into template generation rather then overwriting what the user may have specified?

File: tests/data/dataconverter/readers/ellips/eln_data.yaml [hunk 6]
```diff
@@ -58,9 +58,6 @@ colnames:
 - Delta
 - err.Psi
 - err.Delta
-definition: NXellipsometry
-definition/@url: https://github.com/FAIRmat-NFDI/nexus_definitions/blob/fairmat/contributed_definitions/NXellipsometry.nxdl.xml
-definition/@version: 0.0.2
```
Comment (sanbrock) [inline tests/data/dataconverter/readers/ellips/eln_data.yaml:63]: comes with the change above, that these data are autofilled and not manually configured

File: tests/nexus/test_version.py [hunk 7]
```diff
@@ -0,0 +1,14 @@
...
+from pynxtools import get_nexus_version
+
+
+def test_get_nexus_version():
+    """
+    Tests if we get a version string from nexus definitions
+    """
+    version = get_nexus_version()
+
+    assert version is not None
```
Comment (sanbrock) [inline tests/nexus/test_version.py:13]: should not we test something more specific, like to format of the string returned?

File: MANIFEST.in (hunk 1 above)
Comment (domna) [reply MANIFEST.in:4]: The line was wrongly formatted and no `xsd` files were included in the site-packages. This is different for `recursive-include` (with space in between) and `include` (without space in between).

File: pynxtools/dataconverter/readers/ellips/reader.py (hunk 5 above)
Comment (domna) [reply pynxtools/dataconverter/readers/ellips/reader.py:475]: IMO this *should* overwrite. Even if the user specified it we use this version from pynxtool's submodule for creating the file.

File: pynxtools/_build_wrapper.py (hunk 3 above)
Comment (domna) [reply pynxtools/_build_wrapper.py:16]: mmh... I did not think about this yet as I thought we have it fixed here anyways. But maybe it's a good idea to determine it from the remotes automatically, too (so we don't have to change anything if the repo changes).

Comment (domna) [reply pynxtools/_build_wrapper.py:16]: Let's include this in the TF discussion and then I can add it.

File: pynxtools/dataconverter/helpers.py (hunk 4 above)
Comment (domna) [reply pynxtools/dataconverter/helpers.py:595]: No, this is the attribute in NXroot. The other one is the `/entry/definition` path.

File: tests/data/dataconverter/readers/ellips/eln_data.yaml (hunk 6 above)
Comment (domna) [reply tests/data/dataconverter/readers/ellips/eln_data.yaml:63]: Yes, these are autofilled now.

File: tests/nexus/test_version.py (hunk 7 above)
Comment (sanbrock) [reply tests/nexus/test_version.py:13]: Yes, I agree, but actually I was rather thinking about a regexp check that the format is according expectation

File: pynxtools/dataconverter/readers/ellips/reader.py (hunk 5 above)
Comment (sherjeelshabih) [reply pynxtools/dataconverter/readers/ellips/reader.py:475]: #174 is related to this. I will cross-link this so discussions can be followed in both places. If there is a clear decision here, we can close #174.

I don't have a strong opinion either side. But our project is designed to be used from PyPi. Therefore, overwriting the version to the one that was used to verify is the more coherent way forward. 

If we have an advanced user, they could easily bypass any checks and just use the Writer class with a Template object to write whatever they desired.

**Review by sherjeelshabih — COMMENTED** [review]
I only have suggestion. We should add `nexus-version.txt` to `.gitignore` to make sure one of us doesn't commit that by mistake. From what I understand, that file should not exist outside of the Python wheel. 

Other than that this looks fine to me.

**Comment by domna** [comment]
> I only have suggestion. We should add `nexus-version.txt` to `.gitignore` to make sure one of us doesn't commit that by mistake. From what I understand, that file should not exist outside of the Python wheel.
> 
> Other than that this looks fine to me.

Yes. It's already ignored since we have `*.txt` in `.gitignore.

File: tests/nexus/test_version.py (hunk 7 above)
Comment (domna) [reply tests/nexus/test_version.py:15]: I added this

**Comment by domna** [comment]
@sanbrock Shall we also add an `/@NeXus_repo` field which contains a reference to the repository the version is created from? I think this would align with what we discussed today in the TF.

File: pynxtools/dataconverter/readers/ellips/reader.py (hunk 5 above)
Comment (domna) [reply pynxtools/dataconverter/readers/ellips/reader.py:475]: Yes I would vote for this option from #174:


> The other suggestion is to give an error if the attrib provided by the user is different in value then what the default write function provides. This also makes sense for sensitive attributes like Application Definition or NeXus Version. But if the user really wants to override, we will need to provide a flag for that.

File: pynxtools/_build_wrapper.py (hunk 3 above)
Comment (domna) [reply pynxtools/_build_wrapper.py:16]: Currently, we only get the vcs version here, but we add a repository with the current hash along the actual version. I also added `/@NeXus_repository` to NXroot (which I still need to add to NXroot) for this.

File: pynxtools/dataconverter/helpers.py (hunk 4 above)
Comment (domna) [reply pynxtools/dataconverter/helpers.py:599]: Currently, the writer is able to overwrite this without notice. I would

1. Swap the population of the writer and the NXroot general arguments, that way the general attributes always overwrite the reader ones.
2. Emit a warning if this field is already present there

**Comment by domna** [comment]
Hey @sherjeelshabih, I changed the order in which the NXroot attributes are populated and emit a warning if they are already present. Can you check if this is fine with you as a general approach? Other than that this PR should be good to go now :)

**Comment by sherjeelshabih** [comment]
> Hey @sherjeelshabih, I changed the order in which the NXroot attributes are populated and emit a warning if they are already present. Can you check if this is fine with you as a general approach (we can then continue the discussion in #174 and see if we want to build something on top)? Other than that this PR should be good to go now :)

Hey, Florian. I'm in favor of the approach to warn and update users of the wrong attributes they add. But I see an issue with the implementation right now. 

The update_and_warn function and the add_default_attributes is not really adding these values to the NeXus file. It corrects them in the Template. Then the template does not get written. You should change the order back in the convert.py function.

**Review by sherjeelshabih — APPROVED** [review]

**Comment by domna** [comment]
I added a test for checking the warning message and somehow I cannot get caplog to work. I had this before for the report of undocumented fields with logging. I'm able to see logging errors when I run everything via cli but not in the console or caplog. @sherjeelshabih or @sanbrock Do you know what could be the issue here?

**Pull request merged**